File: basic/cs205_c_cpp_2020s/lab14/rusrc/lab14_02.rs
mod lib;

mod lab14_02 {
    use std::fmt::Display;
    use std::ops::AddAssign;

    use crate::lib::{MAX_COLS, MAX_ROWS};
    use crate::lib::matrix::matrix::Matrix;
    use crate::lib::matrix_str::matrix_str::MatrixS;

    fn use_matrix_template<T: Display + AddAssign + Clone>(m: &mut Matrix<T>,
                                                           array1: &Vec<Vec<T>>,
                                                           array2: &Vec<Vec<T>>) {
        m.set_matrix(&array1);
        println!("Matrix set to first array");
        m.print_matrix();
        m.add_matrix(&array2);
        println!("Matrix incremented by second array");
        m.print_matrix();
    }

    fn use_matrix_template2<T: Display + AddAssign + Clone>(m: &mut Matrix<T>,
                                                            array1: &[[T; MAX_COLS]; MAX_ROWS],
                                                            array2: &[[T; MAX_COLS]; MAX_ROWS], ) {
        m.set_matrix2(&array1);
        println!("Matrix set to first array");
        m.print_matrix();
        m.add_matrix2(&array2);
        println!("Matrix incremented by second array");
        m.print_matrix();
    }

    fn use_matrix_template_str(m: &mut MatrixS,
                               array1: &Vec<Vec<&str>>,
                               array2: &Vec<Vec<&str>>) {
        m.set_matrix(&array1);
        println!("Matrix set to first array");
        m.print_matrix();
        m.add_matrix(&array2);
        println!("Matrix incremented by second array");
        m.print_matrix();
    }

    fn use_matrix_template_str2(m: &mut MatrixS,
                                array1: &[[&str; MAX_COLS]; MAX_ROWS],
                                array2: &[[&str; MAX_COLS]; MAX_ROWS], ) {
        m.set_matrix2(&array1);
        println!("Matrix set to first array");
        m.print_matrix();
        m.add_matrix2(&array2);
        println!("Matrix incremented by second array");
        m.print_matrix();
    }

    pub(crate) fn main() {
        let str1 = vec![
            vec!["Congra", "y", "ar"],
            vec!["alm", "don", "La"]];
        let str2 = vec![
            vec!["tulations", "ou", "e"],
            vec!["ost", "e the", "b!"]];
        const STR3: [[&str; 3]; 2] = [
            ["Congra", "y", "ar", ],
            ["alm", "don", "La", ]];
        const STR4: [[&str; 3]; 2] = [
            ["tulations", "ou", "e"],
            ["ost", "e the", "b!"]];
        let mut string_matrix: MatrixS = MatrixS {
            inside: vec![vec!["".to_string(); MAX_COLS]; MAX_ROWS],
        };
        let num1 = vec![vec![1, 2, 3], vec![4, 5, 6]];
        let num2 = vec![vec![6, 5, 4], vec![3, 2, 1]];
        const NUM3: [[i64; 3]; 2] = [[1, 2, 3], [4, 5, 6]];
        const NUM4: [[i64; 3]; 2] = [[6, 5, 4], [3, 2, 1]];
        let mut int_matrix: Matrix<i64> = Matrix {
            inside: vec![vec![0i64; MAX_COLS]; MAX_ROWS],
        };
        use_matrix_template_str(&mut string_matrix, &str1, &str2);
        use_matrix_template_str2(&mut string_matrix, &STR3, &STR4);
        use_matrix_template(&mut int_matrix, &num1, &num2);
        use_matrix_template2(&mut int_matrix, &NUM3, &NUM4);
    }
}

fn main() {
    lab14_02::main();
}

File: basic/cs205_c_cpp_2020s/lab14/exercise/lab14_02.cpp
// SPDX-License-Identifier: AGPL-3.0-or-later
/*  CS205_C_CPP
    Copyright (C) 2020-2023 nanoseeds
*/
/**
 * @Github: https://github.com/Certseeds
 * @Author: nanoseeds
 * @Date: 2020-05-23 11:16:48
 * @LastEditors  : nanoseeds
 */
#include <iostream>
#include <string>
#include "matrix.h"

using namespace std;

template<typename T1>
void useMatrixTemplate(Matrix<T1> &M, T1 array1[][MAXCOLS], T1 array2[][MAXCOLS]);

int main() {
    string str1[MAXROWS][MAXCOLS]
            {
                    {"Congra", "y",   "ar"},
                    {"alm",    "don", "La"}
            };
    string str2[MAXROWS][MAXCOLS]
            {
                    {"tulations", "ou",    "e"},
                    {"ost",       "e the", "b!"}
            };
    int num1[MAXROWS][MAXCOLS]
            {
                    {1, 2, 3},
                    {4, 5, 6}
            };
    int num2[MAXROWS][MAXCOLS]
            {
                    {6, 5, 4},
                    {3, 2, 1}
            };
    double dou1[MAXROWS][MAXCOLS]
            {
                    {1.6f, 2.5f, 3.4f},
                    {4.3f, 5.2f, 6.1f}
            };
    double dou2[MAXROWS][MAXCOLS]
            {
                    {6.1f, 5.2f, 4.3f},
                    {3.4f, 2.5f, 1.6f}
            };
    Matrix<string> stringMatrix;
    Matrix<int> intMatrix;
    Matrix<double> doubleMatrix;
    cout << "\nDemonstrating with string matrix:" << endl;
    useMatrixTemplate(stringMatrix, str1, str2);
    cout << "\nDemonstrating with int matrix:" << endl;
    useMatrixTemplate(intMatrix, num1, num2);
    cout << "\nDemonstrating with double matrix:" << endl;
    useMatrixTemplate(doubleMatrix, dou1, dou2);
    cout << "\n" << endl;
    return 0;
}

template<typename T1>
void useMatrixTemplate(Matrix<T1> &M, T1 array1[][MAXCOLS], T1 array2[][MAXCOLS]) {
    M.setMatrix(array1);
    cout << "\nMatrix set to first array" << endl;
    M.printMatrix();
    M.addMatrix(array2);
    cout << "\nMatrix incremented by second array" << endl;
    M.printMatrix();
}
